src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -620,7 +620,8 @@ public void stopJenkins() throws Throwable {
         if (!proc.waitFor(60, TimeUnit.SECONDS) ) {
             System.err.println("Jenkins failed to stop within 60 seconds, attempting to kill the Jenkins process");
             proc.destroyForcibly();
-            throw new AssertionError("Jenkins failed to terminate within 60 seconds");
```

Comment:
so to me this is trading one set of flakes for another.

Jenkins could not shut down within the requested time - presumably there is some code that is running / blocking in a `@Terminator`  

Various things could be happening in this state including flushing state to disk, closing databases etc and if the process can not close cleanly these will no longer happen - and then you have failures later on because something that was expected to be persisted was not in fact cleanly persisted.

Did you investigate why Jenkins could not shut down within the 60 seconds?

Reply:
> Did you investigate why Jenkins could not shut down within the 60 seconds?

I'll try to understand it, but locally when running the tests `waitFor` takes no more then 15-20 seconds (to reproduce the flake I had to set timeout to 3 seconds)

Reply:
from the attached logs -> `Shutting down a Jenkins instance that was still starting up`  is likely to cause issues.

Perhaps something is not waiting long enough for Jenkins to start?

Reply:
I like the idea of printing a thread dump for debugging purposes. To get the PID for pre-Java 11 versions of Jenkins you can call `GNUCLibrary.LIBC.getpid()` (guarded by `if (Functions.isGlibcSupported()) {`).

Reply:
From inside the Jenkins JVM you can use https://github.com/jenkinsci/jenkins-test-harness/blob/0038ef9b8a0bb2fec2ef4be67d1124f51db53076/src/main/java/org/jvnet/hudson/test/JenkinsRule.java#L458-L464

Reply:
@jtnord That’s interesting, this log I’ve attached, was collected March 22, then, I assume, the issue with 'port already in use'  was fixed, but Tests continue to be flaky (last failure Jul 23) but the log now is changed and provide much less details. 
[stacktrace.txt](https://github.com/jenkinsci/jenkins-test-harness/files/9196171/stacktrace.txt)
[st_out.txt](https://github.com/jenkinsci/jenkins-test-harness/files/9196173/st_out.txt)
[st_err.txt](https://github.com/jenkinsci/jenkins-test-harness/files/9196175/st_err.txt)
Reading the logs I even don't see that RealJenkinsRule started and ready, I'll try to dig deeper here, maybe I'll understand something

Reply:
it does appear as though Jenkins hasn't started up - I would guess the close is caused due to the teardown due to a test timeout.  
This to me signifies that this is running on a machine that is woefully underpowered, or there is a deadlock at startup, or the timeout is using a wall-clock and there has been a clock reset on the underlying hardware.  Getting the threadDump of the Jenkins process (and any locks) could reveal the smoking gun here.

Reply:
I think this thread can be resolved now?